Serialization

YSaxon opened the pull request:
A couple of methods to serialize the software token, (optionally with a password on the private key).

YSaxon committed:
add serialization methods

bodik commented:
Hi @YSaxon,

thank you for your contribution. I'm just wondering if it's a good idea to bring in `pickle` which is kind of very python specific binary serialization (webauthn itself uses CBOR) and also in some cases it might bring in some security concerns (like insecure deserialization).

I guess that leaving the binary serialization out-of-scope would make the feature (and the library itself) more versatile ? Do you have any thoughts or strong preferences here ?

YSaxon committed:
add test_serialization

YSaxon commented:
@bodik
Good point. I've updated it to use CBOR for byte-serialization instead

bodik commented:
> @bodik Good point. I've updated it to use CBOR for byte-serialization instead

excelent, thank you

bodik committed:
tests: silence deprecation warning for fido2.features.webauthn_json_mapping.enabled

future function would change currently expected data representation

```
  You are using deprecated functionality which will change in the next major version of
  ...
  This changes the keys and values used by the webauthn data classes when accessed using
  the Mapping (dict) interface (eg. user_entity["id"] and the from_dict() methods) to be
  JSON-friendly and align with the current draft of the next WebAuthn Level specification.
  For the most part, this means that binary values (bytes) are represented as URL-safe
  base64 encoded strings instead.
```

also concerns
https://github.com/bodik/soft-webauthn/pull/13

bodik commented:
@YSaxon I've made some changes and squashes. would you review the current PR please ?

bodik commented:
@YSaxon ping, still interested in this ?

eidorb commented:
Thanks, I'm using a similar approach here for now:

https://github.com/eidorb/ubank/blob/9a95de557d5bacd09cfe8633ad8b4bba660ddb56/ubank.py#L174-L198

bodik commented:
> Thanks, I'm using a similar approach here for now:
> 
> https://github.com/eidorb/ubank/blob/9a95de557d5bacd09cfe8633ad8b4bba660ddb56/ubank.py#L174-L198

will the PR implementation be usable as-is in your project so you'd benefit from merge ?

eidorb commented:
Sure, and #15 too 😊 

They both look to do the job and provide tests.

eidorb commented:
Actually, hold off merging for now. I'll take a proper look at these.

bodik commented:
> Actually, hold off merging for now. I'll take a proper look at these.

ok

bodik commented:
@eidorb  so based on #17 discussion we go with this, ok ?

eidorb commented:
The essence here is a desire for serialization. I think we need `SoftWebauthnDevice` to provide a serializable *form* without having to understand its internals. Leave final serialization itself (and encryption) to consumers.

Examples of serialization and encryption without additional dependencies could be given in documentation. 

I think there are two reasonable options:

- `to_dict()` and `from_dict()` give/take serializable forms, probably JSON-serializable forms (bytes and private key serialized to str values)
- implement `dump()`, `dumps()`, `load()`, `loads()` which give/take files/strings

Edit:

- or  `to_bytes()` and `from_bytes()` methods only
- or as-is, with password encryption removed

eidorb commented:
And should support serialization immediately after `SoftWebauthnDevice()` instantiated when `self.private_key` is still None.

Test serialization before calling `.create()`: https://github.com/YSaxon/soft-webauthn/blob/47efdd2b00b4e305ab8a5b92b3086bc6b620e1c9/tests/test_serialization.py#L14

eidorb commented:
Proposal in #18 uses to_dict() and from_dict() -- bytes values are left as-is. I think this is clean.

Example use with CBOR serialization and Fernet password encryption: 

```python
    def dump(self, file: IO[bytes], password: str = ""):
        """Serializes passkey to `file`, encrypted with `password`."""
        # https://cryptography.io/en/latest/fernet/#using-passwords-with-fernet
        fernet = Fernet(derive_key(password))
        file.write(
            fernet.encrypt(
                cbor.encode(
                    {
                        "name": self.name,
                        "hardware_id": self.hardware_id,
                        "device_id": self.device_id,
                        "device_meta": self.device_meta,
                        "username": self.username,
                        "soft_webauthn_device_dict": self.soft_webauthn_device.to_dict(),  # TODO: what will the serializable form of SoftWebauthnDevice be?
                    }
                )
            )
        )
```

https://github.com/eidorb/ubank/blob/c49e2a43746e22f4bba37698d6df614dd8d39185/ubank.py#L264-L281
